ti/instrumentation/utils.go

```diff
@@ -264,6 +267,151 @@ func getChangedFiles(ctx context.Context, workspace string, log *logrus.Logger,
+			message := fmt.Sprintf("%v Files changed after changing bazel file %v; crosses limit set for bazelFileCount", count, file.Name)
+			log.Infoln(message)
+			if javaTestKindOutput != nil {
+				splitName := strings.Split(file.Name, "/")
+				moduleName := splitName[0]
+				moduleList = append(moduleList, moduleName)
+			}
+		} else {
+			// get list of .java files from the output
+			javaFileNames := extractJavaFilesFromQueryOutput(allJavaSrcsOutput)
+			for _, name := range javaFileNames {
+				// to prevent duplicate files
+				if _, exists := uniqueFiles[name]; !exists {
+					changedRes = append(changedRes, ti.File{Status: file.Status, Name: name})
+					uniqueFiles[name] = struct{}{}
+				}
+			}
+		}
+	}
+	return changedRes, moduleList, nil
+}
+
+// takes bazelQueryOutput string and parses it to extract .java files and returns in a list
+func extractJavaFilesFromQueryOutput(bazelOutput []byte) []string {
+	var javaFileNames []string
+	// Outer Regex to extract only srcs
+	pattern1 := `srcs = \[[^\]]+\]`
+	r1 := regexp.MustCompile(pattern1)
+
+	sections := r1.FindAllString(string(bazelOutput), -1)
+	for _, section := range sections {
+		// Inner Regex to match only .java files inside each src
+		pattern2 := `"//[^"]+\.java"`
+		r2 := regexp.MustCompile(pattern2)
+		matches := r2.FindAllString(section, -1)
+		for _, match := range matches {
+			srcFile := strings.Split(strings.Trim(match, `\"`), ":")
+			changedFile := filepath.Join(strings.TrimPrefix(srcFile[0], "//"), srcFile[1])
+			javaFileNames = append(javaFileNames, changedFile)
+		}
+	}
+	return javaFileNames
+}
+
+// takes file name, runs bazel queries extract src globs defined in java rules, viz java_library and java_test
+func getJavaRulesFromBazel(ctx context.Context, file, workspace string) (allSrcs, countAllSrcs, testSrcs []byte, err error) {
+	bazelQueryInput := strings.Replace(file, "/BUILD.bazel", "", -1)
+
+	// query to fetch test src globs from java rule
+	// eg of the query is: bazel query 'kind("java", 980-commons:*)' --output=build  | grep 'srcs ='
+	c1 := fmt.Sprintf("cd %s; %s query 'kind(\"java\", %s:*)' --output=build | grep 'srcs =' ", workspace, bazelCmd, bazelQueryInput)
+	cmdArgs1 := []string{"-c", c1}
+	javaKindOutput, err := execCmdCtx(ctx, "sh", cmdArgs1...).Output()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to run bazel query %v, encountered %v as %v has no java rule", c1, err, file)
+	}
+
+	// query to count the scr files in bazelQueryOutput
+	// eg of the query is: bazel query 'kind("java", 332-ci-manager/app:*)' --output=build  | grep 'srcs =' | grep -o '\.java' | wc -l
+	c2 := fmt.Sprintf("cd %s; %s query 'kind(\"java\", %s:*)' --output=build | grep 'srcs =' | grep -o '\\.java' | wc -l", workspace, bazelCmd, bazelQueryInput)
+	cmdArgs2 := []string{"-c", c2}
+	countQueryOutput, err := execCmdCtx(ctx, "sh", cmdArgs2...).Output()
```

Comment:
Do these commands have a timeout? We should not hang for a long time and give up after some timeout.